### PR TITLE
Add input validation

### LINKED ISSUE #12: Input Validation
Was testing out the new GitHub Releases feature (thanks!), and twice do to some still-waking-up issues managed to misspell or malform the `with`. Would be nice to get some basic validation. Like if `fromGithubReleases` is provided, error if either `repo`, `version`, or `urlTemplate` is not present.

### PR DESCRIPTION
**Pull request by radu-matei**
closes #12

### CODE REVIEW

File: src/configurator.ts [hunk 1]
```diff
@@ -119,6 +120,57 @@ export class Configurator {
 
     core.addPath(toolPath);
   }
+
+  validate() {
+    if (this.name === "") {
+      throw new Error(
+        `the name used to configure the tool in path is required.`
```
Comment (RichiCoder1) [inline src/configurator.ts:127]: Maybe:
```suggestion
        `"name" is required. This is used to set the executable name of the tool`
```

File: src/configurator.ts [hunk 2]
```diff
@@ -119,6 +120,57 @@ export class Configurator {
...
+      );
+    }
+
+    if (!this.fromGitHubReleases && this.url === "") {
+      throw new Error(`url is required if trying to fetch tool directly.`);
```
Comment (RichiCoder1) [inline src/configurator.ts:132]:
```suggestion
      throw new Error(`url is required when using GitHub Releases for versioning`);
```

File: src/configurator.ts [hunk 3]
```diff
@@ -119,6 +120,57 @@ export class Configurator {
...
+    }
+
+    if (!this.fromGitHubReleases && !matchesUrlRegex(this.url)) {
+      throw new Error(`url supplied as input is not a valid URL.`);
+    }
+
+    if (this.fromGitHubReleases && !matchesUrlRegex(this.urlTemplate)) {
+      throw new Error(`urlTemplate supplied as input is not a valid URL.`);
```
Comment (RichiCoder1) [inline src/configurator.ts:140]: Dumb question: have you tested this works with urls with `{{ }}` in them?

File: src/configurator.ts [hunk 4]
```diff
@@ -119,6 +120,57 @@ export class Configurator {
...
+      );
+    }
+
+    if (
+      this.fromGitHubReleases &&
+      getArchiveType(this.urlTemplate) !== ArchiveType.None &&
+      this.pathInArchive === ""
+    ) {
+      throw new Error(
+        `urlTemplate points to an archive, pathInArchive cannot be empty.`
```
Comment (RichiCoder1) [inline src/configurator.ts:158]:
```suggestion
        `pathInArchive is required when urlTemplate points to an archive file.`
```

File: src/configurator.ts [hunk 5]
```diff
@@ -119,6 +120,57 @@ export class Configurator {
...
+    if (this.fromGitHubReleases && !matchesUrlRegex(this.urlTemplate)) {
+      throw new Error(`urlTemplate supplied as input is not a valid URL.`);
+    }
+
+    if (
+      getArchiveType(this.url) !== ArchiveType.None &&
+      this.pathInArchive === ""
+    ) {
+      throw new Error(
+        `url points to an archive, pathInArchive cannot be empty.`
```
Comment (RichiCoder1) [inline src/configurator.ts:148]: For constency sake, I'd put the failing field first in these messages
```suggestion
        `pathInArchive is required when url points to an archive file.`
```

File: src/configurator.ts [hunk 6]
```diff
@@ -119,6 +120,57 @@ export class Configurator {
 
     core.addPath(toolPath);
   }
+
+  validate() {
+    if (this.name === "") {
```
Comment (RichiCoder1) [inline src/configurator.ts:125]: General nit: I think these can be `null` or whitespace under circumstances

**Review by RichiCoder1 — COMMENTED** [review]
Awesome! General coverage looks good, just have some nits about wording and a check

File: src/configurator.ts (hunk 3 above)
Comment (radu-matei) [reply src/configurator.ts:140]: Yeah, tested that manually, plus there is one test case for it (there should probably be more than just one..) - https://github.com/radu-matei/configurator/blob/input-validation/test/action.ts#L143-L163

File: src/configurator.ts (hunk 6 above)
Comment (radu-matei) [reply src/configurator.ts:125]: I think it can be `if(this.name)`, which should check for `null`, `undefined`, and `""`.
Good catch!

**Review by RichiCoder1 — APPROVED** [review]
Nice!

**Pull request merged**